File: serverplugin_lua_proj/src/helpers/serverconfig-reader.lua
local platform = require("src.helpers.platform")
local server_config_reader = {}
server_config_reader.__index = server_config_reader

-- Singleton instance
local instance = nil

-- Constructor
function server_config_reader:new()
    local obj = {
        config_map = {} -- Table to store key-value pairs
    }
    setmetatable(obj, self)
    return obj
end

-- Static method to get the instance
function server_config_reader:get_instance()
    if not instance then
        instance = server_config_reader:new()
        instance:initialize_config();
    end
    return instance
end

-- Method to load and parse the config file
function server_config_reader:load_config(file_path)
    local file = io.open(file_path, "r")
    if not file then
        error("Could not open config file: " .. file_path)
    end

    for line in file:lines() do
        line = line:match("^%s*(.-)%s*$") -- Trim whitespace

        -- Skip empty lines and comments
        if line ~= "" and not line:match("^#") then
            -- Handle inline comments (e.g., key = value #comment)
            local key_value_part = line:match("^(.-)#") or line
            key_value_part = key_value_part:match("^%s*(.-)%s*$") -- Trim again

            -- Split key and value by '='
            local key, raw_value = key_value_part:match("^(.-)%s*=%s*(.-)$")

            if key and raw_value then
                -- Remove surrounding quotes (if any) from the value
                local value = self:strip_quotes(raw_value)
                self.config_map[key] = value
            end
        end
    end

    file:close()
end

-- Method to get all key-value pairs
function server_config_reader:get_all()
    return self.config_map
end

-- Method to get a value by key
function server_config_reader:get_value(key)
    return self.config_map[key] or ""
end

-- Method to get a value with a default fallback
function server_config_reader:get_value_else_default(key, default_value)
    return self.config_map[key] or default_value
end

-- Method to get a value as a number, with a default fallback
function server_config_reader:get_value_as_number(key, default_value)
    local value = self:get_value(key)
    if value ~= "" then
        local parsed_number = tonumber(value)
        if parsed_number then
            return parsed_number
        end
    end
    return default_value
end

-- Helper method to strip quotes from a value
function server_config_reader:strip_quotes(value)
    return value:gsub("^['\"`]", ""):gsub("['\"`]$", "")
end

-- Initialize configuration
function server_config_reader:initialize_config()
    local __dirname = platform.get_workspace_path();
    if os.getenv("SERVER_ENV") == "production" then
        print("Running in production mode")
        self:load_config(platform.join_paths(__dirname, './serverconfig.rel.inf'))
    else
        print("Running in development mode")
        self:load_config(platform.join_paths(__dirname, './serverconfig.rel.inf'))
    end
end

return server_config_reader

File: serverplugin_rust_proj/src/helpers/server_config_reader.rs
#![allow(non_camel_case_types)]  // Disable the warning for non-CamelCase types
use std::collections::HashMap;
use std::fs;

pub struct server_config_reader {
    config_map: HashMap<String, String>,
}

impl server_config_reader {
    // Static method to get the instance
    pub fn get_instance() -> &'static mut server_config_reader {
        static mut INSTANCE: Option<server_config_reader> = None;
        unsafe {
            INSTANCE.get_or_insert_with(|| server_config_reader {
                config_map: HashMap::new(),
            })
        }
    }

    // Method to load and parse the config file
    pub fn load_config(&mut self, file_path: &str) {
        let file_content = fs::read_to_string(file_path).unwrap_or_default();

        for line in file_content.lines() {
            let trimmed_line = line.trim();

            // Skip empty lines and comments
            if trimmed_line.is_empty() || trimmed_line.starts_with('#') {
                continue;
            }

            // Handle inline comments (e.g., key = value #comment)
            let key_value_part = trimmed_line.split('#').next().unwrap_or(trimmed_line);

            // Split key and value by '='
            let parts: Vec<&str> = key_value_part.split('=').map(|part| part.trim()).collect();
            if parts.len() == 2 {
                let key = parts[0].to_string();
                let value = self.strip_quotes(parts[1].to_string());
                self.config_map.insert(key, value);
            }
        }
    }

    // Method to get all key-value pairs as a HashMap
    pub fn get_all(&self) -> HashMap<String, String> {
        self.config_map.clone()
    }

    // Method to get a value by key
    pub fn get_value(&self, key: &str) -> String {
        self.config_map.get(key).cloned().unwrap_or_default()
    }

    pub fn get_value_else_default(&self, key: &str, default_value: &str) -> String {
        self.config_map.get(key).cloned().unwrap_or_else(|| default_value.to_string())
    }

    // Method to get a value as a number, with a default fallback
    pub fn get_value_as_number(&self, key: &str, default_value: u32) -> u32 {
        if let Ok(value) = self.get_value(key).parse::<u32>() {
            value
        } else {
            default_value
        }
    }

    // Helper method to strip quotes from a value
    fn strip_quotes(&self, value: String) -> String {
        value.trim_matches(|c| c == '"' || c == '\'' || c == '`').to_string()
    }
}
